### DISCUSSION TITLE
provide an option to control auto import mode.

### DISCUSSION THREAD

**Discussion by heejaechang** [microsoft/pylance-release#3403] [Enhancement]
Currently, due to perf and UX issue, we made user symbols to mainly show up in ```add import``` and only user symbols already used transitively showed up in ```auto import``` when ```python.analysis.indexing``` is on.

but there are users who just want to make user symbols to show up in ```auto import``` and fine with a few thousands auto import entries showing up in completion time to time.

we should consider adding an option to control the behavior.

related issue - https://github.com/microsoft/pylance-release/issues/3380

**Comment by heejaechang** [comment]
we can also make the option more like

python.analysis.autoImportMode: 
Minimum (default) <= what we have
More <= all user symbols, everything in top ```__init__``` + minimum
Maximum <= scan up to 4 level by default + More

I put 4 level since it looks like most third party libs use less than 4 level nested path. but we can make it bigger or smaller, just remember that number will affect how much files we parse/bind before hand.

**Comment by heejaechang** [comment]
basically, our current behavior is trying to index and show as less as possible in completion. so we can reduce perf hit as much as possible.

the option will provide users an easy way to change that behavior to either little bit more (More) to as much as possible (Maximum) if convenience wins cost from them.

for ```More```, I think cost will be a few seconds more for indexing, and a few hundreds ms more for completion. 
for ```Maximum```, I think cost will be around 10+- seconds more for indexing, and +- 100ms more for completion.

* the cost is linear to number of user files + number of packages. so, cost above is for people who has a few hundred user files and a few popular packages (and many its dependencies) installed.

**Comment by rchiodo** [comment]
I think the Minimum/More/Maximum settings would be hard for people to understand. What about letting them set the limit for how deep we index?

**Comment by heejaechang** [comment]
that's one we already have. the one you reviewed (that Luciana creating a doc in doc website). I think we should still keep that one for users that want to control detail and care about details.

but I believe there are also many people who doesn't care about each specific behavior but want a big knob. like how we expose typeCheckingMode, off, basic, strict. and then expose each rules so that people who want to control in detail and still can do so.

**Comment by rchiodo** [comment]
So the Minimum/More/Maximum would be documented as setting the level? Or does it do more than that?

**Comment by heejaechang** [comment]
by doing more, what do you mean?

**Comment by rchiodo** [comment]
Has some other impact other than setting the indexing level. If the user sets the setting to 'Maximum' is that equivalent to setting the index level to 4? As in it's an alias for the other setting.

**Comment by heejaechang** [comment]
putting user symbols in ```auto import``` or not, currently that doesn't have an option. but except that, user can use existing ```python.analysis.packageIndexDepths``` to achieve what new option provide and more. just more complex and verbose but more fine grained.

**Comment by heejaechang** [comment]
another idea is providing code action to generate requirements file or (other package management file) from actual symbol used by user code so we can reduce number of lib we need to index.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
